wandb/integration/kfp/kfp_patch.py

```diff
@@ -0,0 +1,212 @@
+import inspect
+import itertools
+import textwrap
+from typing import Callable, List, Mapping, Optional
+
+from kfp.components import structures
+from kfp.components._components import _create_task_factory_from_component_spec
+from kfp.components._python_op import _func_to_component_spec
+import wandb
+
+from .wandb_logging import wandb_log
+
+decorator_code = inspect.getsource(wandb_log)
+wandb_logging_extras = f"""
+import typing
+from typing import NamedTuple
+
+import collections
+from collections import namedtuple
+
+import kfp
+from kfp import components
+from kfp.components import InputPath, OutputPath
+
+import wandb
+
+{decorator_code}
+"""
+
+
+def patch(func, module):
+    wandb.patched.setdefault(module.__name__, [])
+    setattr(module, f"orig_{func.__name__}", getattr(module, func.__name__))
+    setattr(module, func.__name__, func)
+    wandb.patched[module.__name__].append([module, func.__name__])
+
+
+def unpatch(module):
+    for module, func in wandb.patched[module.__name__]:
+        setattr(module, func, getattr(module, f"orig_{func}"))
+    wandb.patched[module.__name__] = []
+
+
+def patch_kfp():
+    patch(create_component_from_func, wandb.util.get_module("kfp.components"))
```

Comment:
`get_module` might return nil if someone imports our kfp integration or the kfp team releases a new version that doesn't expose kfp.components.  We should be sure we're being defensive and expecting the worst.

When we detect we don't have what we expect, we should probably just print a `wandb.termerror` and allow our decorators to still be called, but just not do anything.

Essentially, when we're monkey patching we want to avoid crashing a users script if the third party library changes.  It's much better for their code to still work even if they can't get metrics until they downgrade or upgrade.